File: src/main/java/com/gtladd/gtladditions/api/recipe/ChanceParallelLogic.java
package com.gtladd.gtladditions.api.recipe;

import org.gtlcore.gtlcore.api.machine.trait.IRecipeCapabilityMachine;
import org.gtlcore.gtlcore.api.machine.trait.MEPatternRecipeHandlePart;
import org.gtlcore.gtlcore.api.recipe.ingredient.LongIngredient;

import com.gregtechceu.gtceu.api.GTValues;
import com.gregtechceu.gtceu.api.capability.recipe.*;
import com.gregtechceu.gtceu.api.recipe.GTRecipe;
import com.gregtechceu.gtceu.api.recipe.chance.boost.ChanceBoostFunction;
import com.gregtechceu.gtceu.api.recipe.content.Content;
import com.gregtechceu.gtceu.api.recipe.ingredient.FluidIngredient;
import com.gregtechceu.gtceu.api.recipe.ingredient.SizedIngredient;
import com.gregtechceu.gtceu.utils.IngredientEquality;
import com.gregtechceu.gtceu.utils.ItemStackHashStrategy;

import com.lowdragmc.lowdraglib.side.fluid.FluidStack;

import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.Ingredient;

import com.hepdd.gtmthings.common.block.machine.trait.CatalystFluidStackHandler;
import it.unimi.dsi.fastutil.objects.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.Map;
import java.util.function.Predicate;

public class ChanceParallelLogic {

    private static class ContentAmountPair {

        long amount = 0;
        final int chance;
        final int maxChance;

        private ContentAmountPair(int chance, int maxChance) {
            this.chance = chance;
            this.maxChance = maxChance;
        }

        public void addTo(long amount) {
            this.amount += amount;
        }
    }

    public static long getMaxParallel(
                                      IRecipeCapabilityHolder holder,
                                      GTRecipe recipe,
                                      long parallelAmount,
                                      Map<RecipeCapability<?>, Object2IntMap<?>> chanceCaches,
                                      @NotNull ChanceBoostFunction boostFunction,
                                      int baseTier,
                                      int machineTier) {
        for (var cap : recipe.inputs.keySet()) {
            if (cap == ItemRecipeCapability.CAP) {
                parallelAmount = Math.min(parallelAmount, getInputItemParallel(holder, recipe, parallelAmount, chanceCaches.get(ItemRecipeCapability.CAP), boostFunction, baseTier, machineTier));
                if (parallelAmount == 0L) break;
            } else if (cap == FluidRecipeCapability.CAP) {
                parallelAmount = Math.min(parallelAmount, getInputFluidParallel(holder, recipe, parallelAmount, chanceCaches.get(FluidRecipeCapability.CAP), boostFunction, baseTier, machineTier));
                if (parallelAmount == 0L) break;
            }
        }
        return parallelAmount;
    }

    @SuppressWarnings("unchecked")
    static long getInputItemParallel(
                                     IRecipeCapabilityHolder holder,
                                     GTRecipe recipe,
                                     long parallelAmount,
                                     Object2IntMap<?> cache,
                                     @NotNull ChanceBoostFunction boostFunction,
                                     int baseTier,
                                     int machineTier) {
        if (parallelAmount <= 1) return parallelAmount;
        if (!(holder instanceof IRecipeCapabilityMachine machine)) return 1;
        if (machine.getRecipeHandleParts().isEmpty() && machine.getMEPatternRecipeHandleParts().isEmpty()) return 0;

        Object2LongOpenCustomHashMap<Ingredient> guaranteedMap = new Object2LongOpenCustomHashMap<>(IngredientEquality.IngredientHashStrategy.INSTANCE);
        Object2ReferenceOpenCustomHashMap<Ingredient, ContentAmountPair> chanceMap = new Object2ReferenceOpenCustomHashMap<>(IngredientEquality.IngredientHashStrategy.INSTANCE);

        for (Content content : recipe.getInputContents(ItemRecipeCapability.CAP)) {
            if (content.chance <= 0) continue;
            Ingredient ingredient = ItemRecipeCapability.CAP.of(content.content);
            long ingredientCount;
            if (ingredient instanceof LongIngredient longIngredient) {
                ingredientCount = longIngredient.getAmount();
            } else if (ingredient instanceof SizedIngredient sizedIngredient) {
                ingredientCount = sizedIngredient.getAmount();
            } else ingredientCount = 1;

            if (content.chance >= content.maxChance) guaranteedMap.addTo(ingredient, ingredientCount);
            else {
                // Separate guaranteed (chance >= maxChance) from chanced inputs
                // Apply boost function to get the actual chance used in roll
                int boostedChance = boostFunction.getBoostedChance(content, baseTier, machineTier);
                if (boostedChance >= content.maxChance) {
                    guaranteedMap.addTo(ingredient, ingredientCount);
                } else {
                    chanceMap.computeIfAbsent(ingredient, ignored -> new ContentAmountPair(boostedChance, content.maxChance)).addTo(ingredientCount);
                }
            }
        }

        if (guaranteedMap.isEmpty() && chanceMap.isEmpty()) return parallelAmount;

        Object2LongOpenCustomHashMap<ItemStack> ingredientStacks = new Object2LongOpenCustomHashMap<>(ItemStackHashStrategy.comparingAllButCount());

        var handle = machine.getRecipeHandleMap().get(recipe);
        if (handle instanceof MEPatternRecipeHandlePart mePatternRecipeHandlePart) {
            // ME handler
            for (var entry : Object2LongMaps.fastIterable(mePatternRecipeHandlePart.getFirstAvailableMEContentOrEmpty(ItemRecipeCapability.CAP, mePatternRecipeHandlePart.getRecipes2SlotsMap().getValues(recipe)))) {
                ingredientStacks.addTo(entry.getKey(), entry.getLongValue());
            }
        } else if (handle != null) {
            for (var entry : Object2LongMaps.fastIterable(handle.getContent(ItemRecipeCapability.CAP))) {
                ingredientStacks.addTo(entry.getKey(), entry.getLongValue());
            }
        } else {
            // ME handlers, All Active Slots
            for (MEPatternRecipeHandlePart part : machine.getMEPatternRecipeHandleParts()) {
                for (var entry : Object2LongMaps.fastIterable(part.getMEContent(ItemRecipeCapability.CAP))) {
                    ingredientStacks.addTo(entry.getKey(), entry.getLongValue());
                }
            }
            // other handlers
            for (var it : machine.getCapabilities().getOrDefault(IO.IN, Collections.emptyList())) {
                for (var entry : Object2LongMaps.fastIterable(it.getContent(ItemRecipeCapability.CAP))) {
                    ingredientStacks.addTo(entry.getKey(), entry.getLongValue());
                }
            }
        }
        return calculateWithChance(parallelAmount, guaranteedMap, chanceMap, ingredientStacks, (Object2IntMap<Ingredient>) cache);
    }

    @SuppressWarnings("unchecked")
    static long getInputFluidParallel(
                                      IRecipeCapabilityHolder holder,
                                      GTRecipe recipe,
                                      long parallelAmount,
                                      Object2IntMap<?> cache,
                                      @NotNull ChanceBoostFunction boostFunction,
                                      int baseTier,
                                      int machineTier) {
        if (parallelAmount <= 1) return parallelAmount;
        if (!(holder instanceof IRecipeCapabilityMachine machine)) return 1;
        if (machine.getRecipeHandleParts().isEmpty() && machine.getMEPatternRecipeHandleParts().isEmpty()) return 0;

        Object2LongOpenHashMap<FluidIngredient> guaranteedFluidMap = new Object2LongOpenHashMap<>();
        Object2ReferenceOpenHashMap<FluidIngredient, ContentAmountPair> chanceFluidMap = new Object2ReferenceOpenHashMap<>();

        for (Content content : recipe.getInputContents(FluidRecipeCapability.CAP)) {
            if (content.chance <= 0) continue;
            FluidIngredient fluidInput = FluidRecipeCapability.CAP.of(content.content);

            if (content.chance >= content.maxChance) guaranteedFluidMap.addTo(fluidInput, fluidInput.getAmount());
            else {
                // Separate guaranteed from chanced fluid inputs
                // Apply boost function to get the actual chance used in roll
                int boostedChance = boostFunction.getBoostedChance(content, baseTier, machineTier);
                if (boostedChance >= content.maxChance) {
                    guaranteedFluidMap.addTo(fluidInput, fluidInput.getAmount());
                } else {
                    chanceFluidMap.computeIfAbsent(fluidInput, ignored -> new ContentAmountPair(boostedChance, content.maxChance)).addTo(fluidInput.getAmount());
                }
            }
        }

        if (guaranteedFluidMap.isEmpty() && chanceFluidMap.isEmpty()) return parallelAmount;

        Object2LongOpenHashMap<FluidStack> ingredientStacks = new Object2LongOpenHashMap<>();

        var recipeHandle = machine.getRecipeHandleMap().get(recipe);
        if (recipeHandle instanceof MEPatternRecipeHandlePart mePatternRecipeHandlePart) {
            // ME handler
            for (var entry : Object2LongMaps.fastIterable(mePatternRecipeHandlePart.getFirstAvailableMEContentOrEmpty(FluidRecipeCapability.CAP, mePatternRecipeHandlePart.getRecipes2SlotsMap().getValues(recipe)))) {
                ingredientStacks.addTo(entry.getKey(), entry.getLongValue());
            }
        } else if (recipeHandle != null && machine.isDistinct()) {
            for (var entry : Object2LongMaps.fastIterable(recipeHandle.getContent(FluidRecipeCapability.CAP))) {
                ingredientStacks.addTo(entry.getKey(), entry.getLongValue());
            }
        } else {
            for (var container : machine.getCapabilitiesFlat(IO.IN, FluidRecipeCapability.CAP)) {
                if (container instanceof CatalystFluidStackHandler) continue;
                for (var object : container.getContents()) {
                    if (object instanceof FluidStack fs) {
                        ingredientStacks.addTo(fs, fs.getAmount());
                    }
                }
            }
        }
        return calculateWithChance(parallelAmount, guaranteedFluidMap, chanceFluidMap, ingredientStacks, (Object2IntMap<FluidIngredient>) cache);
    }

    private static <I extends Predicate<S>, S> long calculateWithChance(
                                                                        long parallelLimit,
                                                                        Object2LongMap<I> guaranteedMap,
                                                                        Object2ReferenceMap<I, ContentAmountPair> chanceMap,
                                                                        Object2LongMap<S> ingredientStacks,
                                                                        Object2IntMap<I> cache) {
        if (ingredientStacks.isEmpty()) return 0;

        // First, calculate parallel limit for guaranteed inputs (traditional method)
        for (var entry : Object2LongMaps.fastIterable(guaranteedMap)) {
            long needed = entry.getLongValue();

            long available = findAvailable(entry.getKey(), ingredientStacks);
            if (available < needed) return 0;

            parallelLimit = Math.min(parallelLimit, available / needed);
        }

        // For chanced inputs, calculate maximum parallel based on roll simulation
        for (var entry : Object2ReferenceMaps.fastIterable(chanceMap)) {
            ContentAmountPair chanceNeeded = entry.getValue();
            I ingredient = entry.getKey();

            long available = findAvailable(ingredient, ingredientStacks);
            if (available < chanceNeeded.amount) return 0;

            // Get or generate cached chance value
            int cached = getGuaranteedCachedChanceValue(ingredient, chanceNeeded.maxChance, cache);

            // parallel <= ((maxOutputs + 1) * maxChance - 1 - cached) / chance
            // This ensures: floor((parallel * chance + cached) / maxChance) <= maxOutputs
            long maxOutputs = available / chanceNeeded.amount;
            long maxParallelForThis = ((maxOutputs + 1) * chanceNeeded.maxChance - 1 - cached) / chanceNeeded.chance;

            // Ensure non-negative
            if (maxParallelForThis < 0) maxParallelForThis = 0;

            parallelLimit = Math.min(parallelLimit, maxParallelForThis);
        }

        return parallelLimit;
    }

    private static <I extends Predicate<S>, S> long findAvailable(I ingredient, Object2LongMap<S> stacks) {
        for (var it = Object2LongMaps.fastIterator(stacks); it.hasNext();) {
            var input = it.next();
            if (ingredient.test(input.getKey())) {
                it.remove();
                return input.getLongValue();
            }
        }
        return 0;
    }

    private static <I> int getGuaranteedCachedChanceValue(I content, int maxChance, @Nullable Object2IntMap<I> cache) {
        if (cache == null) {
            return GTValues.RNG.nextInt(maxChance);
        }

        if (cache.containsKey(content)) {
            return cache.getInt(content);
        } else {
            // Generate random and save to cache for consistency
            int randomValue = GTValues.RNG.nextInt(maxChance);
            cache.put(content, randomValue);
            return randomValue;
        }
    }
}

File: src/main/java/com/gtladd/gtladditions/api/recipe/ChanceParallelLogic.kt
package com.gtladd.gtladditions.api.recipe

import com.gregtechceu.gtceu.api.GTValues
import com.gregtechceu.gtceu.api.capability.recipe.FluidRecipeCapability
import com.gregtechceu.gtceu.api.capability.recipe.IRecipeCapabilityHolder
import com.gregtechceu.gtceu.api.capability.recipe.ItemRecipeCapability
import com.gregtechceu.gtceu.api.capability.recipe.RecipeCapability
import com.gregtechceu.gtceu.api.recipe.GTRecipe
import com.gregtechceu.gtceu.api.recipe.chance.boost.ChanceBoostFunction
import com.gregtechceu.gtceu.api.recipe.ingredient.FluidIngredient
import com.gregtechceu.gtceu.api.recipe.ingredient.SizedIngredient
import com.gregtechceu.gtceu.utils.FluidStackHashStrategy
import com.gregtechceu.gtceu.utils.IngredientEquality
import com.gregtechceu.gtceu.utils.ItemStackHashStrategy
import com.hepdd.gtmthings.common.block.machine.trait.CatalystFluidStackHandler
import com.lowdragmc.lowdraglib.side.fluid.FluidStack
import it.unimi.dsi.fastutil.objects.*
import net.minecraft.world.item.crafting.Ingredient
import org.gtlcore.gtlcore.api.machine.multiblock.ParallelMachine
import org.gtlcore.gtlcore.api.machine.trait.IRecipeCapabilityMachine
import org.gtlcore.gtlcore.api.machine.trait.MEPatternRecipeHandlePart
import org.gtlcore.gtlcore.api.machine.trait.RecipeHandlePart
import org.gtlcore.gtlcore.api.recipe.ingredient.LongIngredient
import org.gtlcore.gtlcore.utils.NumberUtils
import org.gtlcore.gtlcore.utils.NumberUtils.saturatedAdd
import org.gtlcore.gtlcore.utils.datastructure.Int128
import java.math.BigInteger
import java.util.function.Predicate

object ChanceParallelLogic {

    private class ContentAmountPair(
        val chance: Int,
        val maxChance: Int
    ) {
        var amount: Long = 0
            private set

        fun addTo(amount: Long) {
            this.amount += amount
        }
    }

    fun getMaxParallel(
        holder: ParallelMachine,
        recipe: GTRecipe,
        parallelAmount: Long,
        chanceCaches: Map<RecipeCapability<*>, Object2IntMap<*>>,
        boostFunction: ChanceBoostFunction,
        baseTier: Int,
        machineTier: Int
    ): Long {
        var parallel = parallelAmount
        for (cap in recipe.inputs.keys) {
            if (cap == ItemRecipeCapability.CAP) {
                parallel = minOf(
                    parallel,
                    getInputItemParallel(
                        holder,
                        recipe,
                        parallel,
                        chanceCaches[ItemRecipeCapability.CAP],
                        boostFunction,
                        baseTier,
                        machineTier
                    )
                )
                if (parallel == 0L) break
            } else if (cap == FluidRecipeCapability.CAP) {
                parallel = minOf(
                    parallel,
                    getInputFluidParallel(
                        holder,
                        recipe,
                        parallel,
                        chanceCaches[FluidRecipeCapability.CAP],
                        boostFunction,
                        baseTier,
                        machineTier
                    )
                )
                if (parallel == 0L) break
            }
        }
        return parallel
    }

    @Suppress("UNCHECKED_CAST")
    fun getInputItemParallel(
        holder: ParallelMachine,
        recipe: GTRecipe,
        parallelAmount: Long,
        cache: Object2IntMap<*>?,
        boostFunction: ChanceBoostFunction,
        baseTier: Int,
        machineTier: Int
    ): Long {
        if (parallelAmount <= 1) return parallelAmount
        if (holder !is IRecipeCapabilityMachine) return 1
        if (holder.emptyRecipeHandlePart()) return 0

        val guaranteedMap = Object2LongOpenCustomHashMap(
            IngredientEquality.IngredientHashStrategy.INSTANCE
        )
        val chanceMap = Object2ReferenceOpenCustomHashMap<Ingredient, ContentAmountPair>(
            IngredientEquality.IngredientHashStrategy.INSTANCE
        )
        val confirmMEStock = holder.needConfirmMEStock()

        for (content in recipe.getInputContents(ItemRecipeCapability.CAP)) {
            if (content.chance <= 0) continue
            val ingredient = ItemRecipeCapability.CAP.of(content.content)
            val ingredientCount = when (ingredient) {
                is LongIngredient -> ingredient.actualAmount
                is SizedIngredient -> ingredient.amount.toLong()
                else -> 1L
            }

            if (content.chance >= content.maxChance) {
                guaranteedMap.addTo(ingredient, ingredientCount)
            } else {
                val boostedChance = boostFunction.getBoostedChance(content, baseTier, machineTier)
                if (boostedChance >= content.maxChance) {
                    guaranteedMap.addTo(ingredient, ingredientCount)
                } else {
                    chanceMap.computeIfAbsent(ingredient) {
                        ContentAmountPair(boostedChance, content.maxChance)
                    }.addTo(ingredientCount)
                }
            }
        }

        if (guaranteedMap.isEmpty() && chanceMap.isEmpty()) return parallelAmount

        val ingredientStacks = Object2LongOpenCustomHashMap(
            ItemStackHashStrategy.comparingAllButCount()
        )

        val handle = holder.getActiveRecipeHandle(recipe)
        when (handle) {
            is MEPatternRecipeHandlePart -> {
                for (entry in Object2LongMaps.fastIterable(handle.getMEContent(ItemRecipeCapability.CAP, recipe))) {
                    ingredientStacks.mergeLong(entry.key, entry.longValue, NumberUtils::saturatedAdd)
                }
            }

            is RecipeHandlePart -> {
                for (entry in Object2LongMaps.fastIterable(handle.getSelfContent(ItemRecipeCapability.CAP, confirmMEStock))) {
                    ingredientStacks.mergeLong(entry.key, entry.longValue, NumberUtils::saturatedAdd)
                }
            }

            else -> {
                val sharedRecipeHandlePart = holder.sharedRecipeHandlePart
                if (sharedRecipeHandlePart != null) {
                    for (entry in Object2LongMaps.fastIterable(
                        sharedRecipeHandlePart.getSelfContent(ItemRecipeCapability.CAP, confirmMEStock)
                    )) {
                        ingredientStacks.mergeLong(entry.key, entry.longValue, NumberUtils::saturatedAdd)
                    }
                }
            }
        }

        return calculateWithChance(
            parallelAmount,
            guaranteedMap,
            chanceMap,
            ingredientStacks,
            cache as? Object2IntMap<Ingredient>
        )
    }

    @Suppress("UNCHECKED_CAST")
    fun getInputFluidParallel(
        holder: ParallelMachine,
        recipe: GTRecipe,
        parallelAmount: Long,
        cache: Object2IntMap<*>?,
        boostFunction: ChanceBoostFunction,
        baseTier: Int,
        machineTier: Int
    ): Long {
        if (parallelAmount <= 1) return parallelAmount
        if (holder !is IRecipeCapabilityMachine) return 1
        if (holder.emptyRecipeHandlePart()) return 0

        val guaranteedFluidMap = Object2LongOpenHashMap<FluidIngredient>()
        val chanceFluidMap = Object2ReferenceOpenHashMap<FluidIngredient, ContentAmountPair>()
        val confirmMEStock = holder.needConfirmMEStock()

        for (content in recipe.getInputContents(FluidRecipeCapability.CAP)) {
            if (content.chance <= 0) continue
            val fluidInput = FluidRecipeCapability.CAP.of(content.content)

            if (content.chance >= content.maxChance) {
                guaranteedFluidMap.addTo(fluidInput, fluidInput.amount)
            } else {
                val boostedChance = boostFunction.getBoostedChance(content, baseTier, machineTier)
                if (boostedChance >= content.maxChance) {
                    guaranteedFluidMap.addTo(fluidInput, fluidInput.amount)
                } else {
                    chanceFluidMap.computeIfAbsent(fluidInput) {
                        ContentAmountPair(boostedChance, content.maxChance)
                    }.addTo(fluidInput.amount)
                }
            }
        }

        if (guaranteedFluidMap.isEmpty() && chanceFluidMap.isEmpty()) return parallelAmount

        val ingredientStacks = Object2LongOpenCustomHashMap(FluidStackHashStrategy.comparingAllButAmount())

        val handle = holder.getActiveRecipeHandle(recipe)
        when (handle) {
            is MEPatternRecipeHandlePart -> {
                for (entry in Object2LongMaps.fastIterable(handle.getMEContent(FluidRecipeCapability.CAP, recipe))) {
                    ingredientStacks.mergeLong(entry.key, entry.longValue, NumberUtils::saturatedAdd)
                }
            }

            is RecipeHandlePart -> {
                val content = if (holder.isDistinct) {
                    handle.getSelfContent(FluidRecipeCapability.CAP, confirmMEStock)
                } else {
                    handle.getContentWithShared(FluidRecipeCapability.CAP, confirmMEStock)
                }
                for (entry in Object2LongMaps.fastIterable(content)) {
                    ingredientStacks.mergeLong(entry.key, entry.longValue, NumberUtils::saturatedAdd)
                }
            }

            else -> {
                val sharedRecipeHandlePart = holder.sharedRecipeHandlePart
                if (sharedRecipeHandlePart != null) {
                    for (handler in sharedRecipeHandlePart.getCapability(FluidRecipeCapability.CAP)) {
                        if (handler is CatalystFluidStackHandler) continue
                        for (obj in handler.contents) {
                            if (obj is FluidStack) {
                                ingredientStacks.mergeLong(obj, obj.amount, NumberUtils::saturatedAdd)
                            }
                        }
                    }
                }
            }
        }

        return calculateWithChance(
            parallelAmount,
            guaranteedFluidMap,
            chanceFluidMap,
            ingredientStacks,
            cache as? Object2IntMap<FluidIngredient>
        )
    }

    private fun <I, S> calculateWithChance(
        parallelLimit: Long,
        guaranteedMap: Object2LongMap<I>,
        chanceMap: Object2ReferenceMap<I, ContentAmountPair>,
        ingredientStacks: Object2LongMap<S>,
        cache: Object2IntMap<I>?
    ): Long where I : Predicate<S> {
        if (ingredientStacks.isEmpty()) return 0

        var parallel = parallelLimit

        // First, calculate parallel limit for guaranteed inputs (traditional method)
        for (entry in Object2LongMaps.fastIterable(guaranteedMap)) {
            val needed = entry.longValue
            val available = findAvailable(entry.key, ingredientStacks)
            if (available < needed) return 0

            parallel = minOf(parallel, available / needed)
        }

        // For chanced inputs, calculate maximum parallel based on roll simulation
        for (entry in Object2ReferenceMaps.fastIterable(chanceMap)) {
            val chanceNeeded = entry.value
            val ingredient = entry.key

            val available = findAvailable(ingredient, ingredientStacks)
            if (available < chanceNeeded.amount) return 0

            // Get or generate cached chance value
            val cached = getGuaranteedCachedChanceValue(ingredient, chanceNeeded.maxChance, cache)

            // parallel <= ((maxOutputs + 1) * maxChance - 1 - cached) / chance
            // This ensures: floor((parallel * chance + cached) / maxChance) <= maxOutputs
            val maxOutputs = available / chanceNeeded.amount
            var maxParallelForThis = (Int128(maxOutputs).add(1L)
                .multiply(Int128(chanceNeeded.maxChance.toLong()))
                .subtract(Int128((1 + cached).toLong()))
                .divide(chanceNeeded.chance.toLong()) as Number).toLong()

            // Ensure non-negative
            if (maxParallelForThis < 0) maxParallelForThis = 0

            parallel = minOf(parallel, maxParallelForThis)
        }

        return parallel
    }

    private fun <I, S> findAvailable(
        ingredient: I,
        stacks: Object2LongMap<S>
    ): Long where I : Predicate<S> {
        val it = Object2LongMaps.fastIterator(stacks)
        while (it.hasNext()) {
            val input = it.next()
            if (ingredient.test(input.key)) {
                val value = input.longValue
                it.remove()
                return value
            }
        }
        return 0
    }

    private fun <I> getGuaranteedCachedChanceValue(
        content: I,
        maxChance: Int,
        cache: Object2IntMap<I>?
    ): Int {
        if (cache == null) {
            return GTValues.RNG.nextInt(maxChance)
        }

        return if (cache.containsKey(content)) {
            cache.getInt(content)
        } else {
            // Generate random and save to cache for consistency
            val randomValue = GTValues.RNG.nextInt(maxChance)
            cache.put(content, randomValue)
            randomValue
        }
    }
}
